### ISSUE TITLE
AS923 Dwell Time Problem when TxParamSetupAns is Lost

### ISSUE THREAD

**Issue by psuggate** [Lora-net/LoRaMac-node#614]
**Problem:** An end-device can be sending and listening using `SF11` or `SF12`, while the *Network Server* only sends downlinks using `SF10`, causing none of the downlinks to be received. This can be caused when an uplink containing a `TxParamSetupAns` MAC command is lost, followed by the end-device changing from `SF10` to `SF11` or `SF12`. This results in the NS assuming that the dwell-time limit is still active (because it has not received the `TxParamSetupAns`), whereas the end-device assumes that there is no dwell-time limit.

**Description:** When a device joins our AS923 network (using `SF10`), the first post-join downlink contains a `TxParamSetupReq` MAC command, and this command notifies the device that there is no dwell time limit active for this network. Once the mote sends a corresponding `TxParamSetupAns` response, its state is updated to reflect that no dwell time limit is active.

But if the uplink is lost that contains the `TxParamSetupAns` MAC command, then the NS and end-device will disagree on whether the dwell-time limit is active -- the NS has not seen a `TxParamSetupAns` so the default dwell time limit must be enforced, but the end-device has responded to the NS's `TxParamSetupReq` command, so it sends uplinks as if there is no dwell time limit active.

Therefore, if the end-device changes to `SF11` or `SF12` for uplinks, then it will never hear any responses from the NS, because even though the downlink will use the correct channel, the modulation will be incorrect -- `SF10` for the downlink as the NS has not seen the `TxParamSetupAns` response. And because the end-device will not be listening on `SF10`, it cannot receive any future `TxParamSetupReq` MAC commands (until, eventually, the end-device attempts to rejoin, or performs a soft-reset).

**Fix:** Add the `TxParamSetupAns` command to the list of *"sticky"* MAC commands, so that it will be repeated until the end-device next receives a response from the Network Server (NS). This involves changing just two lines of code, of `LoRaMacCommands.c` of the current `en.i-cube_lrwan.zip` distribution. First, at line 38:

```C
const uint8_t CIDsStickyAnsCmds[] = { MOTE_MAC_DL_CHANNEL_ANS, \
    MOTE_MAC_RX_PARAM_SETUP_ANS, MOTE_MAC_RX_TIMING_SETUP_ANS, \
    MOTE_MAC_TX_PARAM_SETUP_ANS /*  <- NEW */ };
```

which adds `MOTE_MAC_TX_PARAM_SETUP_ANS` to the list of sticky answers, and then add another `case` to the `IsSticky(..)` function, which starts from line 286:

```C
static bool IsSticky( uint8_t cid )
{
    switch( cid )
    {
        case MOTE_MAC_DL_CHANNEL_ANS:
        case MOTE_MAC_RX_PARAM_SETUP_ANS:
        case MOTE_MAC_RX_TIMING_SETUP_ANS:
        case MOTE_MAC_TX_PARAM_SETUP_ANS:  //  <- NEW
            return true;
        default:
            return false;
    }
}
```

**Comment by jamesl-dm** [comment]
Any progress on resolving this? I've had this issue disconnect a few nodes during testing as well, and it looks like it'll need to be addressed by a standard change.

**Comment by reissjason** [comment]
Why is DR2 used to in downlinks to the end-device following an uplink for DR0/DR1?
Rx1 should match the datarate used in the uplink regardless of dwelltime settings. Only if Rx1Offset is used would there ever be a non-matching datarate.
If DR0/DR1 is not allowed then Rx2 at default datarate of DR2 should be used to reach the device.

Would it not be better for the network server to send using DR0/DR1 to get the end-device to stop using DR0/DR1 continually due to incorrect settings? This should not be required since Rx2 is available using an approved datarate.

**Comment by psuggate** [comment]
DR2 is being used for these downlinks because Actility's LRC is strictly enforcing:

> The default JoinReq Data Rate is DR2 (SF10/125KHz), this setting ensures that end-devices are compatible with the 400ms dwell time limitation until the actual dwell time limit is notified to the end-device by the network server via the MAC command “TxParamSetupReq”.

from pp.39, lines 1033-1035 , of v1.0.2 of the regional parameters.

I have talked to Actility about this, and suggested that using `RX2` would fix the problem, without violating the above rule, just as you also suggested. Their response:

>  "this is very unlikely to happen"

even though sending the downlink at a different DR vs the uplink, during RX1, seems out-of-spec to me.

Having `TxParamSetupAns` as a "sticky" MAC command does solve this problem, and is what Actility suggested.

**Comment by jamesl-dm** [comment]
>Would it not be better for the network server to send using DR0/DR1 to get the end-device to stop using DR0/DR1 continually due to incorrect settings?

>even though sending the downlink at a different DR vs the uplink, during RX1, seems out-of-spec to me.

Ironically the DR0/DR1 use isn't due to incorrect settings - it's what the network server asked for. Its just that the network server doesn't realize its request has been actioned already. So it looks up the assumed current state of the node, and sees that the node ought to think the downlink dwell is still on. Then it checks the Rx1Offset table, and sees that an SF12 uplink with downlink dwell on should be replied to at SF10:
![image](https://user-images.githubusercontent.com/28915246/51507718-28e57a00-1e2d-11e9-896e-43fb4e4f028a.png)

To detect and workaround the problem, the server would need to reason that the node shouldn't be sending at SF12 unless it got the TxParamSetupReq, which disables both the uplink and downlink dwell at the same time. If it could work this out, it wouldn't really even need to use Rx2 to reply - it could just update the assumed current state to the intended 'no dwell' settings straight away. But I guess that level of reasoning is just a step too far. It would be better if the standard had a more robust and simple way of resolving this corner case. Making the reply sticky sounds like it would do the job.

**Comment by psuggate** [comment]
Yeah, the Network Server (NS) really should handle this in a smarter way -- the NS can infer from the `FCntUp` value that an uplink was lost as well...

And Acility's LRC (their name for the NS) violates line 1235, that you quote above:

> The RX1 receive window uses the same channel than the preceding uplink.

**Comment by ymasmoudi** [comment]
1/ Regarding the line 1235, LRC (NS) is using the same channel to answer the preceding uplink. 
2/ And as answered previously, It uses correctly DR2 as it didn't receive any response from the sensor that the Dwell time was deactivated.

A missing FCntUp is a sign that an uplink was lost but it doesn't provide any indication whether the device did include a TxParamSetupAns to disable dwell time in that missing frame.

**Comment by mluis1** [comment]
I have pushed this topic to be discussed at the next LoRa-Allaince technical committee meeting which will be held in San Diego the 4th to 6th February 2019.

Currently, we can't modify the code as the sticky `TxParamSetupReq` MAC command answer is not compliant with the current specification.

If the technical committee considers this proposition then the change will most certainly happen trough an ERRATA note for the 1.0.3 specification. Only then we can change this. Nowadays, no NS implementation knows how to handle a sticky `TxParamSetupReq`  MAC command.

**Comment by mluis1** [comment]
Just to inform that the change request has been accepted by the LoRa-Alliance technical committee.

This will be fixed once we start working on LoRaWAN 1.0.4 specification implementation.

I think that we can close this issue.
